### PR TITLE
Mbu/teleportation

### PR DESCRIPTION
**Pull request by mbuchberger1967**
Teleportation/ReferenceImplementation.qs is the relevant change: 
Teleportation Task 4.1. Entangled trio - added comment to solution

### CODE REVIEW

File: Teleportation/ReferenceImplementation.qs [hunk 1]
```diff
@@ -147,6 +147,10 @@ namespace Quantum.Kata.Teleportation {
     operation EntangleThreeQubits_Reference (qAlice : Qubit, qBob : Qubit, qCharlie : Qubit) : Unit
     is Adj {
         
+        // |000⟩ --H⊗H⊗I--> 1/2(|000⟩+|010⟩+|100⟩+|110⟩) --X(c)-> 1/2(|001⟩+|011⟩+|101⟩+|111⟩)
```
Comment (tcNickolas) [inline Teleportation/ReferenceImplementation.qs]: I think this explanation would be a bit more readable if the states were given in comments between the actual gates. This way the learner can also add `DumpMachine` between the gates and verify that the output matches the comment. So the start of the code would look like

    // |000⟩ 

    H(qAlice);

    // 1/sqrt(2) (|0⟩ + |1⟩)|00⟩
    
    H(qBob);
    
    // 1/2 (|000⟩+|010⟩+|100⟩+|110⟩)

and so on.

File: GroversAlgorithm/GroversAlgorithm.ipynb [hunk 2]
```diff
@@ -288,14 +288,16 @@
...
-    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$.\n",
+    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$ $ = \\left(\\begin{matrix}1&0&…&0\\\\0&-1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&-1\\end{matrix}\\right) $\n",
     "\n",
     "<br/>\n",
     "<details>\n",
     "  <summary>Need a hint? Click here </summary>\n",
     "    Note that quantum states are defined up to a global phase.\n",
     "    Thus the state obtained as a result of this operation is the same\n",
-    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state.\n",
+    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state and ignoring the global phase -1.<br>\n",
```
Comment (tcNickolas) [inline GroversAlgorithm/GroversAlgorithm.ipynb]: Looking at my original phrasing with a fresh eye, I'd rephrase it a bit further: "Thus the state obtained as a result of this operation is *equivalent* to the state obtained by flipping the sign of only the $|0...0\\rangle$ basis state (those states differ by a global phase -1)"

File: GroversAlgorithm/GroversAlgorithm.ipynb [hunk 3]
```diff
@@ -288,14 +288,16 @@
...
+    "    $-\\big(2|0...0\\rangle\\langle0...0| - I\\big) = I - 2|0...0\\rangle\\langle0...0| = \\left(\\begin{matrix}-1&0&…&0\\\\0&1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&1\\end{matrix}\\right) $<br>\n",
```
Comment (tcNickolas) [inline GroversAlgorithm/GroversAlgorithm.ipynb]: It should be possible to center this formula to make it a bit easier to read. I would also suggest a neat diagonal dots symbol: 

```suggestion
    "    $$-\\big(2|0...0\\rangle\\langle0...0| - I\\big) = I - 2|0...0\\rangle\\langle0...0| = \\left(\\begin{matrix}-1&0&...&0\\\\0&1&...&0\\\\\\vdots&\\vdots&\\ddots&\\vdots\\\\0&0&...&1\\end{matrix}\\right) $$<br>\n",
```

File: GroversAlgorithm/GroversAlgorithm.ipynb [hunk 4]
```diff
@@ -288,14 +288,16 @@
     "\n",
     " * If the register is in any other basis state, multiply its phase by -1.\n",
     "\n",
-    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$.\n",
+    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$ $ = \\left(\\begin{matrix}1&0&…&0\\\\0&-1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&-1\\end{matrix}\\right) $\n",
```
Comment (tcNickolas) [inline GroversAlgorithm/GroversAlgorithm.ipynb]: A couple of matrix formatting suggestions:
```suggestion
    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$ $ = \\left(\\begin{matrix}1&0&...&0\\\\0&-1&...&0\\\\\\vdots&\\vdots&\\ddots&\\vdots\\\\0&0&...&-1\\end{matrix}\\right) $\n",
```

File: GroversAlgorithm/GroversAlgorithm.ipynb [hunk 5]
```diff
@@ -288,14 +288,16 @@
...
     "\n",
     "<br/>\n",
     "<details>\n",
     "  <summary>Need a hint? Click here </summary>\n",
     "    Note that quantum states are defined up to a global phase.\n",
     "    Thus the state obtained as a result of this operation is the same\n",
-    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state.\n",
+    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state and ignoring the global phase -1.<br>\n",
+    "    $-\\big(2|0...0\\rangle\\langle0...0| - I\\big) = I - 2|0...0\\rangle\\langle0...0| = \\left(\\begin{matrix}-1&0&…&0\\\\0&1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&1\\end{matrix}\\right) $<br>\n",
+    "    This can be implemented as $Controlled Z(|s_0 s_1 \\ldots s_{n-2}\\rangle, |s_{n-1}\\rangle)$ which results in: $|1...10\\rangle \\rightarrow |1...10\\rangle, |1...11\\rangle \\rightarrow -|1...11\\rangle$ due to $Z(|0\\rangle)=|0\\rangle$ and $Z(|1\\rangle)=-|1\\rangle$ if you flip all bits (apply $X$) before and after the operation.  \n",
```
Comment (tcNickolas) [inline GroversAlgorithm/GroversAlgorithm.ipynb]: I would move this part of the hint to the 2nd slot - it's more about *how* to implement this unitary than about *what unitary* to implement, so taking one hint with both parts takes the learner too far, and we want to give them just a little nudge in the right direction. In the 2nd hint you'll be able to give this suggestion, and modify the existing hint to say "Alternatively, you can use the same trick..."

I would also leave out the last detail, suggesting to use Controlled Z but not spelling out that you need to flip all the bits. Something like "Consider the Controlled Z gate, applied with most of the qubits as control and the last qubit as target: (your explanation here up to $Z(|1\\rangle)=-|1\\rangle$)".

**Comment by mbuchberger1967** [comment]
Sounds good!
I’ll do over the weekend and will update the PR

BR
Markus

Von: Mariia Mykhailova <notifications@github.com>
Gesendet: Samstag, 21. März 2020 01:44
An: microsoft/QuantumKatas <QuantumKatas@noreply.github.com>
Cc: Markus <markus@markusbuchberger.de>; Author <author@noreply.github.com>
Betreff: Re: [microsoft/QuantumKatas] Mbu/teleportation (#307)


@tcNickolas requested changes on this pull request.

I did a pass through this PR and left some comments on improving formatting. Thank you, the hint for reflection about zero is a very good idea!

I think it will be best to start with updating this PR and merging it before we proceed reviewing the others. We'll need to resolve conflicts on each of them, since the Grover change is in each PR, so if we merge that change as part of this PR, we'll be able to close #308<https://github.com/microsoft/QuantumKatas/pull/308> and do the conflict resolution for just the two remaining PRs. Sounds good?

________________________________

In Teleportation/ReferenceImplementation.qs<https://github.com/microsoft/QuantumKatas/pull/307#discussion_r395939930>:

> @@ -147,6 +147,10 @@ namespace Quantum.Kata.Teleportation {



     operation EntangleThreeQubits_Reference (qAlice : Qubit, qBob : Qubit, qCharlie : Qubit) : Unit



     is Adj {







+        // |000⟩ --H⊗H⊗I--> 1/2(|000⟩+|010⟩+|100⟩+|110⟩) --X(c)-> 1/2(|001⟩+|011⟩+|101⟩+|111⟩)



I think this explanation would be a bit more readable if the states were given in comments between the actual gates. This way the learner can also add DumpMachine between the gates and verify that the output matches the comment. So the start of the code would look like

// |000⟩







H(qAlice);







// 1/sqrt(2) (|0⟩ + |1⟩)|00⟩







H(qBob);







// 1/2 (|000⟩+|010⟩+|100⟩+|110⟩)



and so on.

________________________________

In GroversAlgorithm/GroversAlgorithm.ipynb<https://github.com/microsoft/QuantumKatas/pull/307#discussion_r395941859>:

>      "\n",



     "<br/>\n",



     "<details>\n",



     "  <summary>Need a hint? Click here </summary>\n",



     "    Note that quantum states are defined up to a global phase.\n",



     "    Thus the state obtained as a result of this operation is the same\n",



-    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state.\n",



+    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state and ignoring the global phase -1.<br>\n",



Looking at my original phrasing with a fresh eye, I'd rephrase it a bit further: "Thus the state obtained as a result of this operation is equivalent to the state obtained by flipping the sign of only the $|0...0\rangle$ basis state (those states differ by a global phase -1)"

________________________________

In GroversAlgorithm/GroversAlgorithm.ipynb<https://github.com/microsoft/QuantumKatas/pull/307#discussion_r395942341>:

>      "\n",



     "<br/>\n",



     "<details>\n",



     "  <summary>Need a hint? Click here </summary>\n",



     "    Note that quantum states are defined up to a global phase.\n",



     "    Thus the state obtained as a result of this operation is the same\n",



-    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state.\n",



+    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state and ignoring the global phase -1.<br>\n",



+    "    $-\\big(2|0...0\\rangle\\langle0...0| - I\\big) = I - 2|0...0\\rangle\\langle0...0| = \\left(\\begin{matrix}-1&0&…&0\\\\0&1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&1\\end{matrix}\\right)<file://left(/begin%7bmatrix%7d-1&0&…&0/0&1&…&0/vdots&/vdots&/vdots&/vdots/0&0&0&1/end%7bmatrix%7d/right)> $<br>\n",



It should be possible to center this formula to make it a bit easier to read. I would also suggest a neat diagonal dots symbol:
⬇️ Suggested change

-    "    $-\\big(2|0...0\\rangle\\langle0...0| - I\\big) = I - 2|0...0\\rangle\\langle0...0| = \\left(\\begin{matrix}-1&0&…&0\\\\0&1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&1\\end{matrix}\\right)<file://left(/begin%7bmatrix%7d-1&0&…&0/0&1&…&0/vdots&/vdots&/vdots&/vdots/0&0&0&1/end%7bmatrix%7d/right)> $<br>\n",



+    "    $$-\\big(2|0...0\\rangle\\langle0...0| - I\\big) = I - 2|0...0\\rangle\\langle0...0| = \\left(\\begin{matrix}-1&0&...&0\\\\0&1&...&0\\\\\\vdots&\\vdots&\\ddots&\\vdots\\\\0&0&...&1\\end{matrix}\\right)<file://left(/begin%7bmatrix%7d-1&0&...&0/0&1&...&0/vdots&/vdots&/ddots&/vdots/0&0&...&1/end%7bmatrix%7d/right)> $$<br>\n",



________________________________

In GroversAlgorithm/GroversAlgorithm.ipynb<https://github.com/microsoft/QuantumKatas/pull/307#discussion_r395942448>:

> @@ -288,14 +288,16 @@



     "\n",



     " * If the register is in any other basis state, multiply its phase by -1.\n",



     "\n",



-    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$.\n",



+    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$ $ = \\left(\\begin{matrix}1&0&…&0\\\\0&-1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&-1\\end{matrix}\\right)<file://left(/begin%7bmatrix%7d1&0&…&0/0&-1&…&0/vdots&/vdots&/vdots&/vdots/0&0&0&-1/end%7bmatrix%7d/right)> $\n",



A couple of matrix formatting suggestions:
⬇️ Suggested change

-    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$ $ = \\left(\\begin{matrix}1&0&…&0\\\\0&-1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&-1\\end{matrix}\\right)<file://left(/begin%7bmatrix%7d1&0&…&0/0&-1&…&0/vdots&/vdots&/vdots&/vdots/0&0&0&-1/end%7bmatrix%7d/right)> $\n",



+    "> This operation implements operator $2|0...0\\rangle\\langle0...0| - I$ $ = \\left(\\begin{matrix}1&0&...&0\\\\0&-1&...&0\\\\\\vdots&\\vdots&\\ddots&\\vdots\\\\0&0&...&-1\\end{matrix}\\right)<file://left(/begin%7bmatrix%7d1&0&...&0/0&-1&...&0/vdots&/vdots&/ddots&/vdots/0&0&...&-1/end%7bmatrix%7d/right)> $\n",



________________________________

In GroversAlgorithm/GroversAlgorithm.ipynb<https://github.com/microsoft/QuantumKatas/pull/307#discussion_r395943075>:

>      "\n",



     "<br/>\n",



     "<details>\n",



     "  <summary>Need a hint? Click here </summary>\n",



     "    Note that quantum states are defined up to a global phase.\n",



     "    Thus the state obtained as a result of this operation is the same\n",



-    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state.\n",



+    "    as the state obtained by flipping the sign of only the $|0...0\\rangle$ state and ignoring the global phase -1.<br>\n",



+    "    $-\\big(2|0...0\\rangle\\langle0...0| - I\\big) = I - 2|0...0\\rangle\\langle0...0| = \\left(\\begin{matrix}-1&0&…&0\\\\0&1&…&0\\\\\\vdots&\\vdots&\\vdots&\\vdots\\\\0&0&0&1\\end{matrix}\\right)<file://left(/begin%7bmatrix%7d-1&0&…&0/0&1&…&0/vdots&/vdots&/vdots&/vdots/0&0&0&1/end%7bmatrix%7d/right)> $<br>\n",



+    "    This can be implemented as $Controlled Z(|s_0 s_1 \\ldots<file://ldots> s_{n-2}\\rangle, |s_{n-1}\\rangle)$ which results in: $|1...10\\rangle \\rightarrow<file://rightarrow> |1...10\\rangle, |1...11\\rangle \\rightarrow<file://rightarrow> -|1...11\\rangle$ due to $Z(|0\\rangle)=|0\\rangle$ and $Z(|1\\rangle)=-|1\\rangle$ if you flip all bits (apply $X$) before and after the operation.  \n",



I would move this part of the hint to the 2nd slot - it's more about how to implement this unitary than about what unitary to implement, so taking one hint with both parts takes the learner too far, and we want to give them just a little nudge in the right direction. In the 2nd hint you'll be able to give this suggestion, and modify the existing hint to say "Alternatively, you can use the same trick..."

I would also leave out the last detail, suggesting to use Controlled Z but not spelling out that you need to flip all the bits. Something like "Consider the Controlled Z gate, applied with most of the qubits as control and the last qubit as target: (your explanation here up to $Z(|1\rangle)=-|1\rangle$)".

**Comment by mbuchberger1967** [comment]
I applied your changes. 
BR Markus
